src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java

```diff
@@ -0,0 +1,235 @@
+import com.puppycrawl.tools.checkstyle.utils.TokenUtils;
+
+/**
+ * Generates xpath queries. Xpath queries are generated based on received
+ * {@code DetailAst} element, line number and column number.
+ *
+ * <p>
+ *     Example class
+ * </p>
+ * <pre>
+ * public class Main {
+ *
+ *     public String sayHello(String name) {
+ *         return "Hello, " + name;
+ *     }
+ * }
+ * </pre>
+ *
+ * <p>
+ *     Following expression returns list of queries. Each query is the string representing full
+ *     path to the node inside Xpath tree, whose line number is 3 and column number is 4.
+ * </p>
+ * <pre>
+ *     new XpathQueryGenerator(rootAst, 3, 4).generate();
+ * </pre>
+ *
+ * <p>
+ *     Result list
+ * </p>
+ * <ul>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']
+ *     </li>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']/MODIFIERS
+ *     </li>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']/MODIFIERS/LITERAL_PUBLIC
+ *     </li>
+ * </ul>
+ *
+ * @author Timur Tibeyev.
+ */
+public class XpathQueryGenerator {
+    /** The root ast. */
+    private final DetailAST rootAst;
+    /** The line number of the element for which the query should be generated. */
+    private final int lineNumber;
+    /** The column number of the element for which the query should be generated. */
+    private final int columnNumber;
+
+    /**
+     * Creates a new {@code XpathQueryGenerator} instance.
+     *
+     * @param rootAst root ast
+     * @param lineNumber line number of the element for which the query should be generated
+     * @param columnNumber column number of the element for which the query should be generated
+     */
+    public XpathQueryGenerator(DetailAST rootAst, int lineNumber, int columnNumber) {
+        this.rootAst = rootAst;
+        this.lineNumber = lineNumber;
+        this.columnNumber = columnNumber;
+    }
+
+    /**
+     * Returns list of xpath queries of nodes, matching line and column number.
+     * This approach uses DetailAST traversal. DetailAST means detail abstract syntax tree.
+     * @return list of xpath queries of nodes, matching line and column number
+     */
+    public List<String> generate() {
+        return getMatchingAstElements(rootAst, lineNumber, columnNumber)
+            .stream()
+            .map(XpathQueryGenerator::generateXpathQuery)
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Returns child {@code DetailAst} element of the given root,
+     * which has child element with token type equals to {@link TokenTypes#IDENT}.
+     * @param root {@code DetailAST} root ast
+     * @return child {@code DetailAst} element of the given root
+     */
+    private static DetailAST findChildWithIdent(DetailAST root) {
+        return TokenUtils.findFirstTokenByPredicate(root,
+            cur -> {
+                return cur.findFirstToken(TokenTypes.IDENT) != null;
```

Comment:
But I had violation that there must be `{}` after `->`, I checked other usages of `TokenUtils.findFirstTokenByPredicate` in source files and every where there are curly brackets with return statement